simpleXRPL docs (#224)
ripple/opensource.ripple.com: docs/simpleXRPL/get-started.md

---
seo:
    description: Install simpleXRPL, initialize the client, connect a custodian, and run your first XRP Ledger operation.
labels:
  - SDKs
---

# Get Started

This guide takes you from install to your first on-ledger operation in three steps: **initialize the client**, **set up a custodian**, and **run a vertical operation**. The examples target the XRPL Testnet with a local signer so you can run them as-is, then swap in a production custodian when you're ready.

## Prerequisites

- **Node.js >= 20.19.** simpleXRPL is Node-targeted and not intended to run in the browser.
- Install the package:

```sh
npm install simplexrpl
```

## 1. Initialize the client

> `SimpleXRPL.init(...)` is the single entry point — you never construct the client with `new`. It binds one or more already-authenticated signing backends (the **connectors**) to a network and builds the account index.

Reviewer: Can we standardize how we refer to it? it should be simpleXRPL OR simplexrpl (per the code) but not SimpleXRPL @pdp2121

## 2. Set up a custodian

The **connector** determines how operations run and who holds the keys. simpleXRPL ships three, each constructed and authenticated on its own, then handed to `init`:

- **`LocalSigner`** — self-custody; keys held in-process. For development and testing.
- **`RippleCustody`** — routes through Ripple Custody. For production.
- **`PalisadeCustody`** — routes through Palisade. For production.

The local signer in Step 1 is enough for development. For production, construct a custodian connector and bind it instead of (or alongside) the local one:

```ts
import {
  LocalSigner,
  PalisadeCustody,
  RippleCustody,
  SimpleXRPL,
} from 'simplexrpl'

// Ripple Custody — authenticates with an intent-author key and operates within
// one Custody domain. `fromEnv` reads the `RIPPLE_CUSTODY_*` variables.
const rippleCustody = await RippleCustody.fromEnv({
  primary: process.env.RIPPLE_CUSTODY_PRIMARY ?? '',
})

// Palisade — authenticates via OAuth client credentials and acts on a
// specific vault/wallet.

> const palisade = await PalisadeCustody.create({

Reviewer: just confirming that it's correct that we have no url for ripple custody but we do for palisade? @pdp2121

Reviewer 2: Yes that's correct afaik. Cant tag Cybele here to confirm

Author: Just added her to review team, tagging @cybele-ripple

  primary: {
    vaultId: process.env.PALISADE_VAULT_ID ?? '',
    walletId: process.env.PALISADE_WALLET_ID ?? '',
  },
  // Enable the raw sign-only fallback for transactors the custodian has no
  // native operation for. Off by default.
  allowRawSigning: false,
})

const client = await SimpleXRPL.init({
  rippledUrl: 'wss://s.altnet.rippletest.net:51233',
  signers: [rippleCustody, palisade, LocalSigner.fromEnv()],
  primarySigner: rippleCustody,
})
```

Read credentials from your environment or secrets manager — never hard-code keys. Once bound, every vertical verb works the same regardless of which connector owns the account: the SDK routes each write to the custodian that holds it.

{% admonition type="info" name="Note" %}
Whether an operation runs through a custodian's **native** path or the **raw-signing fallback** is decided per operation, and fallback is off unless you opt in. See [Operation Execution](index.md#operation-execution) and the [Connector Routing](references/connector-routing.md) table.
{% /admonition %}

See [Connect to custodians](tutorials/connect-custodians.md) for the full per-connector setup.

## 3. Run a vertical operation

> Operations are grouped into domain-specific **verticals** — `xrp`, `token`, `iou`, `credential`, `domain`, and `account` — reached off the client. Each verb reads as business intent rather than protocol mechanics. Here's the simplest one, a native XRP payment:

Reviewer: I think calling it a "payment" defeats the purpose a bit. Can we call it a transfer of XRP from one account to another? and we can specify after that something like (an XRP payment in strict rippled terms or something)

Author: Updated to refer to it as transfers. Will make sure payments aren't referenced anywhere else.

---
seo:
    description: Install simpleXRPL, initialize the client, connect a custodian, discover your accounts, check routing, and send your first XRP Ledger payment.
labels:
  - simpleXRPL
  - SDK
---

# Get Started

> This tutorial takes you through the basics of sending your first operation on the XRP Ledger with simpleXRPL.

Reviewer 3: ```suggestion
This tutorial takes you through the basics of sending your first operation on the XRP Ledger with **simpleXRPL**.
```

## Steps

### 1. Install dependencies

```sh
npm install simplexrpl
```

### 2. Initialize the client

`SimpleXRPL.init(...)` is the single entry point — you never construct the client with `new`. It binds one or more already-authenticated signing backends (the **connectors**) to a network and builds the account index.

{% code-snippet file="/_code-samples/simplexrpl/getStarted.ts" language="ts" before="// --- Discover your accounts ---" /%}

- **`primarySigner`** is the default backend for verbs called without an explicit account. It defaults to `signers[0]`, so you only set it when you bind more than one connector.
- With **no `signers`**, the client is read-only: reads work, but write verbs throw `NoSignerError` until a signer is added.
- Bind an account at runtime (for example, a freshly created wallet) with `client.registerLocalAccount(seed)`.

See [Client and initialization](references/index.md#client-and-initialization) for the full configuration reference.

### 3. Connect a custodian

> The local signer above is enough for development. For production, construct a custodian connector and pass it to `init`'s `signers` — in place of, or alongside, the local one. simpleXRPL ships **Ripple Custody** and **Palisade**, each constructed and authenticated on its own:

Reviewer 3: ```suggestion
The local signer above is enough for development. For production, construct a custodian connector and pass it to  `init`'s `signers`, either in place of or alongside the local one. simpleXRPL ships **Ripple Custody** and **Palisade**, each constructed and authenticated on its own:
```

Author: This has been updated to a step about constructing a connector

{% code-snippet file="/_code-samples/simplexrpl/getStarted.ts" language="ts" from="// --- Connect a custodian (production) ---" /%}

> Read credentials from your environment or secrets manager — never hard-code keys. Once bound, every vertical verb works the same regardless of which connector owns the account: the SDK routes each write to the custodian that holds it.

Reviewer 3: ```suggestion
Read credentials from your environment or secrets manager. Never hard-code keys. Once bound, every vertical verb works the same regardless of which connector owns the account: the SDK routes each write to the custodian that holds it.
```

{% admonition type="info" name="Note" %}

> Whether an operation runs through a custodian's **native** path or the **raw-signing fallback** is decided per operation; the fallback is off by default and enabled per connector via `allowRawSigning`. See [Operation Execution](index.md#operation-execution) and the [Connector Routing](references/connectors/connector-routing.md) table.

Reviewer 3: ```suggestion
Whether an operation runs through a custodian's **native** path or the **raw-signing fallback** is decided per operation. The fallback is off by default and is enabled per connector via `allowRawSigning`. See [Operation Execution](index.md#operation-execution) and the [Connector Routing](references/connectors/connector-routing.md) table.
```

{% /admonition %}

### 4. Discover your accounts

> Connectors discover their accounts at init; the client merges them into a single index keyed by r-address. List them, resolve the primary, and read on-chain state — a read needs no signer:

Reviewer 3: ```suggestion
Connectors discover their accounts at `init`, and the client merges them into a single index keyed by r-address. List them, resolve the primary, and read on-chain state. A read doesn't need a signer:
```

{% code-snippet file="/_code-samples/simplexrpl/getStarted.ts" language="ts" from="// --- Discover your accounts ---" before="// --- Check how operations route ---" /%}

### 5. Check how operations route

Before you submit, ask how a given transactor would route for an account — signed locally, through a custodian's native operation, via the raw sign-only fallback, or rejected:

{% code-snippet file="/_code-samples/simplexrpl/getStarted.ts" language="ts" from="// --- Check how operations route ---" before="// --- Send a payment ---" /%}

### 6. Send a payment

> Operations are grouped into domain-specific **verticals** — `xrp`, `token`, `iou`, `credential`, `domain`, and `account` — reached off the client. Each verb uses the primary account by default; target a different bound account with `from`. Here's a native XRP payment:

Reviewer 3: ```suggestion
Operations are grouped into domain-specific **verticals**, reached off the client: `xrp`, `token`, `iou`, `credential`, `domain`, and `account`. Each verb uses the primary account by default. To target a different bound account, use `from`. Here's a native XRP payment:
```
